File: Neetcode/Trees/Python/SameTree.py
# link: https://leetcode.com/problems/same-tree/
# difficulty: easy
# date solved: oct 2, 2023
# solution
# this is just a traversal and at each time we want to check
# that the trees are the same. first we look if they are both
# none which means we are at the bottom and they are the same tree,
# and return true in this case. the next case we check if either p 
# or q are none, this means that the first case failed so either one is none
# or they both are not none, return false if this case passes. last, we know 
# they are both not none so we check the values to see if they are the same,
# then recurse on the left and the rigth subtrees

class treenode:
    def __init__(self, val=0, left=none, right=none):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p, q):
        if p == None and q == None:
            return True
        elif p == None or q == None:
            return False
        elif p.val != q.val:
            return False
        return self.isSameTree(p.right, q.right) and self.isSameTree(p.left, q.left)

File: Neetcode/Trees/CPP/SameTree.cpp
// Link: https://leetcode.com/problems/same-tree/
// Difficulty: Easy
// Date solved: May 12, 2023
/*
Solution: 
This problem is just checking if two binary trees are the exact same. Our base case is when BOTH p and q are null which means that we can return true.
We also have to check if either p or q are null because if one is null and one is not, then the would pass the first check and we want to return false since
these are clearly not the same. If both nodes are not null, we want to then check the value and return false if they are not the same. We do this for the left and 
the right side and then return these two values AND
*/

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(p == NULL && q == NULL) {
            return true;
        } else if (p == NULL || q == NULL) {
            return false;
        } else if (p->val != q->val) {
            return false;
        }

        bool isSameLeft = isSameTree(p->left, q->left);
        bool isSameRight = isSameTree(p->right, q->right);

        return isSameLeft && isSameRight;
    }
};

int main() {
    Solution sol;
    return 0;
}